### ISSUE TITLE
How to download new version course

### ISSUE THREAD

**Issue by hyhy01** [coursera-dl/coursera-dl#508]
https://www.coursera.org/learn/crypto

Downloading class: crypto
Starting new HTTPS connection (1): class.coursera.org
Starting new HTTPS connection (1): www.coursera.org
Logged in on coursera.org.
Starting new HTTPS connection (1): class.coursera.org
Found authentication cookies.
Downloading about page from: https://api.coursera.org/api/catalog.v1/courses?fie
lds=largeIcon,photo,previewLink,shortDescription,smallIcon,smallIconHover,univer
sityLogo,universityLogoSt,video,videoId,aboutTheCourse,targetAudience,faq,course
Syllabus,courseFormat,suggestedReadings,instructor,estimatedClassWorkload,aboutT
heInstructor,recommendedBackground,subtitleLanguagesCsv&q=search&query=crypto
Starting new HTTPS connection (1): api.coursera.org
Subtitle unavailable in specified language
Downloaded https://class.coursera.org/crypto/lecture/index (1511786 bytes)
Found 0 sections and 0 lectures on this page
The cookies file may be invalid, please re-run with the `--clear-cache` option.

**Comment by balta2ar** [comment]
Are you sure you are enrolled to the old version of the course and you have accepted the term (you do that upon your first visit of the course)? I can download this course just fine.

**Comment by balta2ar** [comment]
Did you try this? https://github.com/coursera-dl/coursera-dl/issues/300#issuecomment-68147829

**Comment by balta2ar** [comment]
@hyhy01 I'm afraid you will need to hack the script a bit.

In this line: https://github.com/coursera-dl/coursera-dl/blob/master/coursera/coursera_dl.py#L1293 please replace `download_class` with `download_on_demand_class` and it will work.

The reason is that there are both versions available and the script tries the old-style version first. There is no switch to prefer new-style first. With this regard this course is very unique as other courses usually have different names on the new on-demand platform.

I'm not sure we need to fix that as old classes are going to be shut down the end of this month. @rbrito What do you think?

**Comment by rbrito** [comment]
Hi, people.

On Fri, Jun 17, 2016 at 3:15 PM, Yuri Bochkarev
<notifications@github.com> wrote:
> I'm not sure we need to fix that as old classes are going to be shut down
> the end of this month. @rbrito What do you think?

Yes, 100% agreed. The next release after the end of this month, we can
release a new version of the script with the code removed. Fewer lines
to debug/test/etc.

It seems that we have really close aesthetic decisions regarding code, BTW. :)

-- 
Rogério Brito : rbrito@{ime.usp.br,gmail.com} : GPG key 4096R/BCFCAAAA
http://cynic.cc/blog/ : github.com/rbrito : profiles.google.com/rbrito
DebianQA: http://qa.debian.org/developer.php?login=rbrito%40ime.usp.br

**Comment by dhawalhshah** [comment]
I see this error when trying to download Learning How To Learn. I wonder if this is the same issue? 
```
Downloading class: learning-how-to-learn
Starting new HTTPS connection (1): class.coursera.org
404 Client Error: Not Found for url: https://class.coursera.org/learning-how-to-learn
Starting new HTTPS connection (1): www.coursera.org
Logged in on coursera.org.
Downloaded https://www.coursera.org/api/opencourse.v1/course/learning-how-to-learn (37570 bytes)
Parsing syllabus of on-demand course. This may take some time, be patient ...
Gathering video URLs for video_id <mGhggl_JEeSfFCIACxcQxA>.
Proceeding with download of resolution 540p of <mGhggl_JEeSfFCIACxcQxA>.
Gathering subtitle URLs for video_id <mGhggl_JEeSfFCIACxcQxA>.
Gathering transcript URLs for video_id <mGhggl_JEeSfFCIACxcQxA>.
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/coursera/cookies.py", line 91, in login
    r.raise_for_status()
  File "/usr/local/lib/python3.4/dist-packages/requests/models.py", line 844, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: https://class.coursera.org/learning-how-to-learn

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.4/dist-packages/coursera/coursera_dl.py", line 1031, in download_class
    return download_old_style_class(args, class_name)
  File "/usr/local/lib/python3.4/dist-packages/coursera/coursera_dl.py", line 912, in download_old_style_class
    username=args.username, password=args.password)
  File "/usr/local/lib/python3.4/dist-packages/coursera/cookies.py", line 345, in get_cookies_for_class
    get_authentication_cookies(session, class_name, username, password)
  File "/usr/local/lib/python3.4/dist-packages/coursera/cookies.py", line 166, in get_authentication_cookies
    login(session, username, password, class_name=class_name)
  File "/usr/local/lib/python3.4/dist-packages/coursera/cookies.py", line 94, in login
    raise ClassNotFound(class_name)
coursera.cookies.ClassNotFound: learning-how-to-learn

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/bin/coursera-dl", line 9, in <module>
    load_entry_point('coursera==0.5.2', 'console_scripts', 'coursera-dl')()
  File "/usr/local/lib/python3.4/dist-packages/coursera/coursera_dl.py", line 1056, in main
    if download_class(args, class_name):
  File "/usr/local/lib/python3.4/dist-packages/coursera/coursera_dl.py", line 1034, in download_class
    return download_on_demand_class(args, class_name)
  File "/usr/local/lib/python3.4/dist-packages/coursera/coursera_dl.py", line 987, in download_on_demand_class
    args.video_resolution)
  File "/usr/local/lib/python3.4/dist-packages/coursera/coursera_dl.py", line 331, in parse_on_demand_syllabus
    video_resolution, assets)
  File "/usr/local/lib/python3.4/dist-packages/coursera/api.py", line 73, in extract_links_from_lecture
    links, self._extract_links_from_lecture_assets(assets))
  File "/usr/local/lib/python3.4/dist-packages/coursera/api.py", line 127, in _extract_links_from_lecture_assets
    for open_course_asset_id in self._get_open_course_asset_ids(asset_id):
  File "/usr/local/lib/python3.4/dist-packages/coursera/api.py", line 153, in _get_open_course_asset_ids
    for element in dom['elements']]
  File "/usr/local/lib/python3.4/dist-packages/coursera/api.py", line 153, in <listcomp>
    for element in dom['elements']]
KeyError: 'assetId'

```

**Comment by balta2ar** [comment]
@dhawalhshah You're using old "coursera" package from pip. Please either use the latest version from github of use "coursera-dl" (with "-di" suffix) package from pip (at least version 0.6.0).

**Comment by khaledoun** [comment]
Hi Everybody,
Despite I replace download_class with download_on_demand_class , but I still had the same problem with this course and also with othr courses : https://www.coursera.org/learn/objective-c

~/coursera/my-coursera/coursera-dl$ sudo coursera-dl -u *******@***** -p ********  objective-c
coursera_dl version 0.6.1
Downloading class: objective-c
Starting new HTTPS connection (1): class.coursera.org
Starting new HTTPS connection (1): www.coursera.org
Logged in on coursera.org.
Starting new HTTPS connection (1): class.coursera.org
Could not authenticate: Did not find necessary cookies.


Can you help me please?

**Comment by balta2ar** [comment]
@khaledoun Are you sure you did it in the right place? I don't think it will keep mentioning `class.coursera.org` if you substitute the function.

**Comment by khaledoun** [comment]
@balta2ar thank you for your reply. yes I am sure that doing in the right place. what I need to change?
(my-coursera) khalid@khalid-Latitude-E6400:~/coursera/my-coursera/coursera-dl$ ls
appveyor             CHANGELOG.md          coursera-dl.bat    MANIFEST.in                setup.py
appveyor.yml      CONTRIBUTING.md      deploy                 README.md                 tox.ini
assets                 coursera                         fabfile.py             requirements-dev.txt
AUTHORS.md    coursera-dl                     LICENSE             requirements.txt

**Comment by balta2ar** [comment]
@khaledoun In fact, you could try using the latest version from github according to this instruction: https://github.com/coursera-dl/coursera-dl#alternative-installation-method-for-unix-systems. Old-courses support is already removed there.

**Comment by laike9m** [comment]
@balta2ar I'm on Mac using virtualenv and latest version of coursera-dl. But it still gives me `Could not authenticate: Did not find necessary cookies.` error.

**Comment by balta2ar** [comment]
@laike9m Did you clone in from github? What's the course name?

**Comment by laike9m** [comment]
No I didn't clone. Coursera name is `software-security` which I've enrolled. Do I need to clone and install from source?

**Comment by balta2ar** [comment]
@laike9m Until new version is released, yes.

**Comment by laike9m** [comment]
It works. Thx a lot 👍 
Does coursera-dl supports skipping downloading specific file(s)?

**Comment by balta2ar** [comment]
Yes, consider the following arguments:
```
--formats FILE_FORMATS
--ignore-formats IGNORE_FORMATS
--section_filter SECTION_FILTER
--lecture_filter LECTURE_FILTER
--resource_filter RESOURCE_FILTER
```
Most likely you need `--ignore-formats IGNORE_FORMATS`.

**Comment by abhigenie92** [comment]
I am getting the same issue. I am using the latest version on pip.
```
abhishek Algos1 $ coursera-dl -u abhigenie@live.com -p XXXXXX algorithm-design-analysis
coursera_dl version 0.6.1
Downloading class: algorithm-design-analysis
Starting new HTTPS connection (1): class.coursera.org
Starting new HTTPS connection (1): www.coursera.org
Logged in on coursera.org.
Starting new HTTPS connection (1): class.coursera.org
Could not authenticate: Did not find necessary cookies.
```

**Comment by k88albr** [comment]
Thanks @balta2ar as of 10/1, still needed to change the script.  Several hoops to jump through to get this all working on OS X El Capitan, so putting in current steps for anyone else.
First installed python 3.52 from https://www.python.org/downloads/

Mostly followed @gokul-uf post here: https://github.com/coursera-dl/coursera-dl/issues/475

```
mkdir mycoursera; cd mycoursera
sudo pip install virtualenv
virtualenv -p /usr/local/bin/python3 coursera
cd coursera/
./bin/pip install coursera
cd ../
./coursera/bin/coursera-dl -u <email> -p <pass> course
```

Which failed with the cookies error.  

Wasn't sure where to find the file to change and found it here:

_./mycoursera/coursera/lib/python3.5/site-packages/coursera/coursera_dl.py_

And in this case, the call at line 1056 'download_class'

`            if download_class(args, class_name):
`

I changed to

  `          if download_on_demand_class(args, class_name):
`

Then it all worked, and I was able to dl the class, thank you so much!
